torch/export/_trace.py

```diff
@@ -232,19 +232,37 @@ def lookup_fake(x):
     return fake_args, fake_kwargs, fake_mode
 
 
-def _replace_param_buffer_names(param_buffer_table, sig):
+def _replace_param_buffer_names(param_buffer_table, sig, mod, original_state_dict):
+    # Some moules have state_dict hooks, which modifies the state_dict to be different
+    # from the module hierarchy. We find a matching between the state_dict with hook
+    # and the state_dict hook.
+
+    state_dict_keys_map = {}
+
+    if mod._state_dict_hooks or mod._state_dict_pre_hooks:
+        state_dict_without_hook = mod.state_dict(keep_vars=True, use_hooks=False)
+        param_id_to_state_dict_key_map = {
+            id(value): key for key, value in state_dict_without_hook.items()
+        }
+        for key, value in original_state_dict.items():
+            value_id = id(value)
+            if value_id in param_id_to_state_dict_key_map:
+                state_dict_keys_map[param_id_to_state_dict_key_map[value_id]] = key
+
     for spec in sig.input_specs:
         if spec.kind in (
             InputKind.PARAMETER,
             InputKind.BUFFER,
         ):
             spec.target = param_buffer_table[spec.target]
+            spec.target = state_dict_keys_map.get(spec.target, spec.target)
     for spec in sig.output_specs:
         if spec.kind in (
             OutputKind.BUFFER_MUTATION,
             OutputKind.GRADIENT_TO_PARAMETER,
         ):
             spec.target = param_buffer_table[spec.target]
+            spec.target = state_dict_keys_map.get(spec.target, spec.target)
```

Comment:
This looks like for the FQN, we store the post-processed, state_dict key, not necessarily the path where the attribute exists. From the discussion I'm not 100% sure what the semantics are, and if this breaks unflattening? Do you know if export() + unflatten() works for the test case you added?

cc: @angelayi

Reply:
discussed in char, summarize the discussion here:

we can modify the verifier to match against the state_dict without the hook

- add a test for unflattening
- add a context manager, so in export, module’s state_dict hook is removed. So now in verifiers, we are matching against the state_dict without hook. Effectively, we ignore state_dict_hooks in export.

One caveat is then for any model with state_dict hook, one won’t be able to  interchange between export_program.state_dict() & mod.state_dict(). 

 export_program.state_dict() still works with it self, but you can’t load a model’s state_dict to exported_program, or vice versa.